### ISSUE TITLE
RFC(rest): ✨ the abstraction of ratelimit data storage ✨

### ISSUE THREAD

**Issue by vladfrangu** [discordjs/discord.js#8125]
### Discussed in https://github.com/discordjs/discord.js/discussions/8124

<div type='discussions-op-text'>

<sup>Originally posted by **didinele** June 19, 2022</sup>
*Preface: this RFC (Request For Comments) is meant to gather feedback and opinions for the feature set and some implementation details for an abstraction of how `@discordjs/rest` stores ratelimit data.*

The goal is to give the end-user a way to control how ratelimit data is stored, enabling things like using Redis to share ratelimits across REST instances (i.e. across different services/processes).

___

# 1. The public API
`RESTOptions` would take a new `store?: IRestStore` parameter, meaning you'd now be able to do `new REST({ store: new YourCustomStore() });`

Currently, I imagine stores as simple key-value storage with a few methods:
```ts
// name up for discussion
interface IRateLimitStore {
  has(key: string): Awaitable<boolean>;
  get(key: string): Awaitable<RateLimitState>;
  set(key: string, data: RateLimitState): Awaitable<void>;
  delete(key: string): Awaitable<void>;
}
```

where `type Awaitable<T> = T | Promise<T>;` and 
```ts
interface RateLimitState {
  reset: number;
  remaining: number;
  limit: number;
  refs: number;
}
```

# 2. Implementation details

First off, what do we use as the key in our little store? The same key we use to store our `IHandler`s. This will come in handy in a bit.
https://github.com/discordjs/discord.js/blob/358c3f4a46638fc818e65fcc113d84818bb68c10/packages/rest/src/lib/RequestManager.ts#L308

From here on out, each `IHandler` implementation can get its state using `await this.manager.rateLimitStore.get(this.id);` 

And lastly, we need to figure out how to sweep this stuff! The main issue is that doing an interval driven sweep of the store via some `lastUsed` property wouldn't be very efficient, especially since, in this case, every `REST` instance using our particular store (e.g. a Redis server) would be redundantly all trying to sweep.

This is where our `refs` property comes into play. Let's assume we have 2 microservices, each one with its own `REST` instance, all using a Redis store.

Now, let's assume a route `/some/route` that has never been hit before. Our first microservice tries to fire up the request, and it eventually needs to create a handler, which we can see being done here:
https://github.com/discordjs/discord.js/blob/358c3f4a46638fc818e65fcc113d84818bb68c10/packages/rest/src/lib/RequestManager.ts#L328-L335

When we create our `SequentialHandler`, we would call `IRateLimitStore#set` with `queue.id` and `{ ...initialState, refs: 1 }`, where the inital state is the current jazz (e.g. `limit: -1`).

Notice how we set our `refs` property to 1. Once our second microservice tries to make the same request, we'll check if state already exists using `IRateLimitStore#has` - which it does - after which we'll simply increment the `refs` property to `2`.

Finally, a few hours later the sweeper will start getting to our handlers:
https://github.com/discordjs/discord.js/blob/358c3f4a46638fc818e65fcc113d84818bb68c10/packages/rest/src/lib/RequestManager.ts#L256-L266

When this happens, we'll query the state for the handler and decrement the ref count by 1. If it's dropped to 0, it means there's currently no active handlers, and therefore the state can be dropped, leading to a `IRateLimitStore#delete` call.</div>

**Comment by n1c00o** [comment]
The implementation seems good however there are some behavioural problems here.

While reference counting avoids the overhead of interval-based task queues, it requires a lot more of locking, especially in distributed systems (multi-threading or micro-services), this can badly affects the speed of systems.
I think it might be worth benchmarking both solutions to determine which solution is more adapted to the scenario.

As of the sweeping determination, the idea of garbage-collecting inactive handlers at specified interval is a good solution in runtime-based systems, it has major flaws which need to be considered:

1. It creates computation spikes at intervals, which will impact the performance of downstream users depending on the environment (the issue is similar to [Discord's problem with Go's garbage collector](https://discord.com/blog/why-discord-is-switching-from-go-to-rust) causing spikes, from my understanding).
2. What is the expected behaviour in case the external store (in your example, a shared Redis cache) in unavailable (network failure, ...). And if the REST service is restarted, does that mean it will remember pending rate limits?
3. At scale, if multiple instances of REST tries to simultaneously sweep handlers, there is a risk of work duplication resulting on unexpected behaviour and/or useless bandwidth consumption. How to avoid it without locking (which would badly slows the system)?

**Comment by didinele** [comment]
In the past, with the appearance of the discord.js proxy I viewed this as the next logical step, for the point at which the proxy server needed scaling.

Ultimately though, the more one thinks about this, the more the benefits of scalable REST diminish.

I'm not familiar with any algorithm or approach that can safely guarantee each REST instance is fully aware of the current context (and can therefore prevent 429s) without constantly ensuring sync on every data operation.

In simpler terms, this is the classic instance of trying to multithread an application, but you find that at almost all times, all but one thread is actually doing work, while the rest wait for their turn on a mutex.

This sort of parallelism isn't my cup of tea, so if someone smarter than me has a proposal, I'm willing to look into it.

For now, the only idea I have is potentially scaling by putting a work balancer in front and deciding which REST instance to use depending on the route, removing the need for any data sharing. This is quite complex, as it likely requires your work balancer to dynamically analyze traffic and decide which instance is responsible for what on the fly. It also means there's still a pretty hard limit; a particularly busy instance can't be scaled up. If this is even worth it, it's very different to this RFC either way, so I will be closing this.

As a closing thought, I think horizontal scaling isn't the way to go here ultimately. If our REST can't keep up anymore, run a proxy like Twilight's and route your traffic through that. Something written in a proper, fast language, should never fail considering Discord's relatively low global ratelimits to begin with.
